lib/driver.js

```diff
@@ -1075,10 +1079,131 @@ export class XCUITestDriver extends BaseDriver {
   }
 
   /**
-   *
-   * @param {string} cmd
-   * @param {...any} args
-   * @returns {Promise<any>}
+   * Checks if an error is related to WDA connection issues
+   * @param {Error|null|undefined} err - The error to check
+   * @returns {boolean} - True if the error is a WDA connection error
+   */
+  isWdaConnectionError(err) {
+    if (!err) {
+      return false;
+    }
+    const errorMessage = err.message || '';
+    const CONNECTION_ERRORS = ['ECONNREFUSED', 'socket hang up'];
+
+    return CONNECTION_ERRORS.some((errStr) => errorMessage.includes(errStr));
+  }
+
+  /**
+   * Reconnects to WDA within a specified timeout
+   * @returns {Promise<object>} - The WDA status
+   * @throws {Error} - If WDA does not become reachable within the timeout
+   */
+  async reconnectWda() {
+    const timeout = this.opts.wdaConnectionTimeout || 15000;
+    const interval = this.opts.wdaStartupRetryInterval || 1000;
+    const startTime = Date.now();
+
+    this.log.info(`Attempting to reconnect to WDA with ${timeout}ms timeout`);
+
+    while (Date.now() - startTime < timeout) {
+      try {
+        const isRunning = await this.wda.isRunning();
+        if (isRunning) {
+          this.log.info(`WDA is now reachable with status: ${isRunning}}`);
+          return isRunning;
+        } else {
+          this.log.debug(`WDA still not reachable, status: ${isRunning}. Retrying in ${interval}ms...`);
+          await this.delay(interval);
+        }
+      } catch (err) {
+        this.log.debug(`WDA still not reachable: ${err.message}. Retrying in ${interval}ms...`);
+        await this.delay(interval);
+      }
+    }
+
+    throw new Error(`Failed to reconnect to WDA within ${timeout}ms`);
+  }
+
+  /**
+   * Creates a promise that resolves after the specified time
+   * @param {number} ms - Delay time in milliseconds
+   * @returns {Promise<void>}
+   */
+  async delay(ms) {
+    return new Promise((resolve) => setTimeout(resolve, ms));
+  }
+
+  /**
+   * Reinitializes the WDA session
+   * @returns {Promise<string>} - The session ID
+   * @throws {Error} - If session reinitialization fails
+   */
+  async reinitializeSession() {
+    this.log.info('Reinitializing WDA session mapping...');
+    const originalSessionId = this.sessionId;
+
+    try {
+      const isRunning = await this.wda.isRunning();
+
+      if (!isRunning) {
+        this.log.info('WDA process is not running. Restarting WDA process...');
+        if (this.wda?.fullyStarted) {
+          try {
+            await this.wda.quit();
+          } catch (quitErr) {
+            this.log.warn(`Error quitting stale WDA: ${quitErr.message}`);
+          }
+        }
+        await this.startWda();
+      } else {
+        this.log.info('WDA process is still running');
+      }
+
+      await this.startWdaSession(this.opts.bundleId, this.opts.processArguments);
```

Comment:
If `this.startWda();` was called in the above line in `if (!isRunning) {`, this `this.startWdaSession` occurs twice because `this.startWda();` also does the same steps. Is it expected?